packages/data-broker/src/DataBroker/DataBroker.ts

```diff
@@ -306,19 +178,23 @@ export class DataBroker {
     options: PullOptions,
   ): Promise<AnalyticResults> {
     const dataElements = await fetchDataElements(this.models, dataElementCodes);
-    const validatedOptions = setOrganisationUnitCodes(options);
+    const organisationUnitCodes = getOrganisationUnitCodes(options);
+    const pulls = await this.getPulls(dataElements, organisationUnitCodes);
+    const allowedOrgUnits = await fetchAllowedOrgUnitsForDataElements(
+      this.models,
+      dataElements,
+      this.context.accessPolicy,
+      organisationUnitCodes,
+    );
 
-    const pulls = await this.getPulls(dataElements, validatedOptions.organisationUnitCodes);
-    const type = this.getDataSourceTypes().DATA_ELEMENT;
     const nestedResults = await Promise.all(
-      pulls.map(({ dataSources: dataSourcesForThisPull, serviceType, dataServiceMapping }) => {
-        return this.pullForServiceAndType(
-          dataSourcesForThisPull,
-          validatedOptions,
-          type,
-          serviceType,
+      pulls.map(({ dataSources: dataElementsForThisPull, serviceType, dataServiceMapping }) => {
+        const service = this.createService(serviceType);
+        return service.pull(dataElementsForThisPull, this.getDataSourceTypes().DATA_ELEMENT, {
```

Comment:
The refactoring of this PR will allow us to change the `Service` interface to something like the following:

```
return service.pullAnalytics(dataElementsForThisPull, options);
```